### PR TITLE
fix: Ensure that popover focus lock works as expected

### PR DESCRIPTION
**Pull request by gethinwebster**
### Description

Prevents some unexpected side-effects from popover focus lock, specifically when the popover
is placed inside a focusable element (i.e. if any parent has `tabindex` set).

Recreation of the problem before fix: https://codesandbox.io/s/recursing-tharp-7ybx7d?file=/src/App.js

### How has this been tested?

Local testing & new tests.

### Documentation changes

[*Do the changes include any API documentation changes?*]
- [ ] _Yes, this change contains documentation changes._
- [X] _No._

### Related Links

AWSUI-19360

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- [ ] _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](CONTRIBUTING.md#public-apis)._
- [ ] _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](CONTRIBUTING.md#browsers-support)._
- [ ] _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- [ ] _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- [ ] _Changes are covered with new/existing unit tests?_
- [ ] _Changes are covered with new/existing integration tests?_
</details>

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by fa7ad — COMMENTED** [review]
> [popover-placement-test-ts-L66](https://github.com/cloudscape-design/components/pull/437/files#file-src-popover-__integ__-popover-placement-test-ts-L66)

Is FocusLock affecting the CSS somehow? the integration test for placement seems to be failing?

**Comment by gethinwebster** [comment]
The popover dismiss button was being focused while the popover was still repositioning itself, therefore it was offscreen & caused the browser to reset scroll. That in turn affected the positioning logic.

**Review by fa7ad — DISMISSED** [review]

File: src/popover/body.tsx [hunk 1]
```diff
@@ -50,6 +49,10 @@ export default function PopoverBody({
     [onDismiss]
   );
 
+  // used to ensure that we only autoFocus the dismiss button
+  // once the popover has been repositioned
+  useEffect(() => setRendered(true), []);
```
Comment (pan-kot) [inline src/popover/body.tsx:54]: I guess the rendered is set to true when the popover is opened, not repositioned?

Comment (pan-kot) [reply src/popover/body.tsx:54]: Why is it necessary to defer the auto-focus?

File: src/popover/internal.tsx [hunk 2]
```diff
@@ -57,6 +57,8 @@ function InternalPopover(
   const clickFrameId = useRef<number | null>(null);
 
   const [visible, setVisible] = useState(false);
+  const [wasVisible, setWasVisible] = useState(false);
+  const [dismissedViaDocumentClick, setDismissedViaDocumentClick] = useState(false);
```
Comment (pan-kot) [inline src/popover/internal.tsx:61]: it is possible to simplify the state by replacing `visible` with some `displayState` union, like:

```
type DisplayState = { type: 'hidden' } | { type: 'visible' } | { type: 'dismissed', withClick: boolean }
```

File: src/popover/body.tsx (hunk 1 above)
Comment (gethinwebster) [reply src/popover/body.tsx:54]: Yes, but that allows the parent PopoverContainer to complete its initial positioning logic before we try to focus the popover content. Otherwise, the PopoverBody may be outside the visible screen (that's how some of the positioning logic is done), and therefore can cause unwanted scrolling when it's focused.

File: src/internal/components/chart-popover/index.tsx [hunk 3]
```diff
@@ -77,6 +78,12 @@ export default function ChartPopover({
     };
   }, [container, onDismiss]);
 
+  const dismissButtonRefCallback = useCallback((element: ButtonProps.Ref) => {
+    if (element) {
+      element.focus();
```
Comment (pan-kot) [inline src/internal/components/chart-popover/index.tsx:83]: why is this necessary?

File: src/popover/internal.tsx [hunk 4]
```diff
@@ -99,6 +102,19 @@ function InternalPopover(
...
   }, []);
 
+  useEffect(() => {
+    if (
+      visible === false &&
+      wasVisible === true &&
+      document.activeElement === document.body &&
+      !dismissedViaDocumentClick
+    ) {
+      triggerRef.current?.focus();
```
Comment (pan-kot) [inline src/popover/internal.tsx:112]: It is possible to do it the moment dismiss happens? E.g. when the dismiss button is pressed or Esc keypress is intercepted?

Having this as a side-effect often leads to tricky issues.

Comment (pan-kot) [reply src/popover/internal.tsx:112]: Or maybe it is possible to change onDismiss:

```
const onDismiss = useCallback(() => {
    setVisible(false);
    if (shouldFocusTrigger()) {
      triggerRef.current?.focus();
    }
}, []);
```

The shouldFocusTrigger part can check the last interaction -- if it was mouse-down or TAB keypress -- ignore, otherwise -- focus.

File: src/popover/body.tsx (hunk 1 above)
Comment (pan-kot) [reply src/popover/body.tsx:54]: Is there some deferring logic in the library focus-lock? We can maybe update our implementation with e.g. adding a 0-timeout here: https://github.com/cloudscape-design/components/blob/main/src/internal/components/focus-lock/index.tsx#L31

File: src/internal/components/chart-popover/index.tsx (hunk 3 above)
Comment (gethinwebster) [reply src/internal/components/chart-popover/index.tsx:83]: To focus the dismiss button when the popover is already open, but is changed into "locked" mode by clicking/pressing enter on a specific point in the chart. The focus-lock library seemed to handle this magically somehow, but doing it more explicitly seems to make sense. Perhaps it would make sense to move this logic elsewhere though, I'll look into this

Comment (gethinwebster) [reply src/internal/components/chart-popover/index.tsx:83]: I've looked, and don't think it's reasonable/sensible to move this logic elsewhere: it would require passing refs through multiple layers of components, and adding the logic separately for each chart type.

File: src/popover/internal.tsx (hunk 4 above)
Comment (gethinwebster) [reply src/popover/internal.tsx:112]: Seems a good simplification, taking a look

File: src/popover/body.tsx (hunk 1 above)
Comment (gethinwebster) [reply src/popover/body.tsx:54]: I've not been able to get to the bottom of why it worked before, but it seems like this shouldn't be the responsibility of the FocusLock component, but rather of the component that uses it, to ensure it's ready before requesting focus lock.

File: src/popover/internal.tsx [hunk 5]
```diff
@@ -58,20 +58,29 @@ function InternalPopover(
...
-  const onTriggerKeyDown = useCallback((event: React.KeyboardEvent) => {
-    if (event.keyCode === KeyCode.tab || event.keyCode === KeyCode.escape) {
-      setVisible(false);
-    }
-  }, []);
+  const onTriggerKeyDown = useCallback(
+    (event: React.KeyboardEvent) => {
+      if (event.keyCode === KeyCode.tab || event.keyCode === KeyCode.escape) {
+        setVisible(false);
+        focusTrigger();
```
Comment (pan-kot) [inline src/popover/internal.tsx:79]: I think the trigger should not be focused when received TAB keypress. The only exception is when expandToViewport is true -- in that case we can't focus elsewhere so focusing on the trigger is reasonable.

Comment (gethinwebster) [reply src/popover/internal.tsx:79]: good catch, this handler doesn't need to refocus ever in fact, as the trigger is by definition already focused...

File: src/mixed-line-bar-chart/__tests__/mixed-chart.test.tsx [hunk 6]
```diff
@@ -928,8 +928,6 @@ describe('Details popover', () => {
 
     wrapper.findChart()!.fireEvent(new MouseEvent('mousedown', { bubbles: true }));
 
-    expect(wrapper.findDetailPopover()!.findDismissButton()!.getElement()).toHaveFocus();
```
Comment (gethinwebster) [inline src/mixed-line-bar-chart/__tests__/mixed-chart.test.tsx:931]: this is now tested in integ tests (see above) as I was unable to get it to work stably here (and testing such interactions in unit tests anyway can be quite difficult/problematic)

File: src/internal/components/chart-popover/index.tsx (hunk 3 above)
Comment (gethinwebster) [reply src/internal/components/chart-popover/index.tsx]: refactored to make this hopefully all a lot clearer

File: src/popover/body.tsx (hunk 1 above)
Comment (gethinwebster) [reply src/popover/body.tsx]: went with an approach of focusing without scrolling instead, so that the exact timing is not important

File: src/annotation-context/annotation/annotation-popover.tsx [hunk 7]
```diff
@@ -78,16 +77,9 @@ export function AnnotationPopover({
...
-      if (element) {
-        element.focus({ preventScroll: true });
-        // Falls back to alignTop on IE11
-        scrollElementIntoView(trackRef.current ?? undefined);
-      }
-    },
-    [trackRef]
-  );
+  useEffect(() => {
+    scrollElementIntoView(trackRef.current ?? undefined);
```
Comment (pan-kot) [inline src/annotation-context/annotation/annotation-popover.tsx:81]: When this is fired?

From the code it looks like the scroll is made only once - when the annotation popover is rendered (but not when it is opened or reopened)?

File: src/mixed-line-bar-chart/chart-container.tsx [hunk 8]
```diff
@@ -284,12 +284,13 @@ export default function ChartContainer<T extends ChartDataTypes>({
...
   }, [highlightedX, showPopover]);
 
-  const onSVGMouseDown = () => {
+  const onSVGMouseDown = (e: React.MouseEvent<SVGSVGElement, MouseEvent>) => {
     if (isPopoverOpen) {
       if (isPopoverPinned) {
         dismissPopover();
       } else {
         pinPopover();
+        e.preventDefault();
```
Comment (pan-kot) [inline src/mixed-line-bar-chart/chart-container.tsx:293]: why prevent default is needed?

Comment (gethinwebster) [reply src/mixed-line-bar-chart/chart-container.tsx:293]: Otherwise the click event continues, causing focus to be moved to the clicked element (the chart itself)

File: src/popover/body.tsx [hunk 9]
```diff
@@ -34,12 +31,12 @@ export default function PopoverBody({
   children,
   onDismiss,
   variant,
-  returnFocus = true,
   overflowVisible,
-  dismissButtonRef,
   className,
 }: PopoverBodyProps) {
   const labelledById = useUniqueId('awsui-popover-');
+  const [dismissButtonFocused, setDismissButtonFocused] = useState(false);
```
Comment (pan-kot) [inline src/popover/body.tsx:38]: As this state is only used inside an effect it can be implemented with a ref to avoid unnecessary re-render:

```
const dismissButtonFocusedRef = useRef(false)

useEffect(() => {
  if (showDismissButton && !dismissButtonFocusedRef.current) {
    dismissButtonRef.current?.focus({ preventScroll: true });
  }
  dismissButtonFocusedRef.current = showDismissButton
}, [showDismissButton])
```

File: src/annotation-context/annotation/annotation-popover.tsx (hunk 7 above)
Comment (pan-kot) [reply src/annotation-context/annotation/annotation-popover.tsx:81]: Oh, I see - thanks. I think it would make sense to explain this in the comments above the use-effect.

File: src/mixed-line-bar-chart/chart-container.tsx (hunk 8 above)
Comment (pan-kot) [reply src/mixed-line-bar-chart/chart-container.tsx:293]: Do we need to apply similar change to area-chart and pie-chart then?

Comment (gethinwebster) [reply src/mixed-line-bar-chart/chart-container.tsx:293]: yes actually, seems these weren't covered by any tests, added the change & tests for these two as well

**Review by pan-kot — APPROVED** [review]
LGTM!
